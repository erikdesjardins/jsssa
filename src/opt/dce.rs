use crate::ir;
use crate::ir::fold::Folder;

#[derive(Default)]
pub struct Dce {
    dropping_after_jump: bool,
}

impl Folder for Dce {
    type Output = Option<ir::Stmt>;

    fn wrap_scope<R>(&mut self, enter: impl FnOnce(&mut Self) -> R) -> R {
        let r = enter(self);
        // stop dropping when we leave the current scope
        self.dropping_after_jump = false;
        r
    }

    fn fold(&mut self, stmt: ir::Stmt) -> Self::Output {
        if self.dropping_after_jump {
            return None;
        }

        match stmt {
            ir::Stmt::Expr {
                ref target,
                ref expr,
            } if target.used().is_never() => match expr {
                ir::Expr::Function { .. }
                | ir::Expr::Bool { .. }
                | ir::Expr::Number { .. }
                | ir::Expr::String { .. }
                | ir::Expr::Null
                | ir::Expr::Undefined
                | ir::Expr::This
                | ir::Expr::Read { .. }
                | ir::Expr::ReadMutable { .. }
                | ir::Expr::Array { .. }
                | ir::Expr::Object { .. }
                | ir::Expr::RegExp { .. }
                | ir::Expr::Unary { .. }
                | ir::Expr::Binary { .. }
                | ir::Expr::CurrentFunction
                | ir::Expr::Argument { .. } => None,
                ir::Expr::ReadGlobal { .. }
                | ir::Expr::ReadMember { .. }
                | ir::Expr::Delete { .. }
                | ir::Expr::Yield { .. }
                | ir::Expr::Await { .. }
                | ir::Expr::Call { .. } => Some(stmt),
            },
            ir::Stmt::DeclareMutable { ref target, val: _ } if target.used().is_never() => None,
            ir::Stmt::Return { .. }
            | ir::Stmt::Throw { .. }
            | ir::Stmt::Break
            | ir::Stmt::Continue => {
                self.dropping_after_jump = true;
                Some(stmt)
            }
            ir::Stmt::IfElse {
                cond: _,
                ref cons,
                ref alt,
            } if cons.0.is_empty() && alt.0.is_empty() => None,
            ir::Stmt::Try {
                ref body,
                catch: _,
                ref finally,
            } if body.0.is_empty() && finally.0.is_empty() => None,
            _ => Some(stmt),
        }
    }
}
