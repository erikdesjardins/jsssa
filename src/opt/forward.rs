use std::collections::HashMap;
use std::iter;

use crate::ir;
use crate::ir::traverse::{visit_with, Folder, ScopeTy};

/// Forward `ir::Expr::Read` to the source SSA ref.
///
/// Does not profit from multiple passes.
/// Does not profit from DCE running first; may create opportunities for DCE.
#[derive(Default)]
pub struct Reads {
    ssa_remappings: HashMap<ir::WeakRef<ir::Ssa>, ir::Ref<ir::Ssa>>,
}

impl Reads {
    fn remap(&self, ref_: ir::Ref<ir::Ssa>) -> ir::Ref<ir::Ssa> {
        match self.ssa_remappings.get(&ref_.weak()) {
            Some(orig_ref) => orig_ref.clone(),
            None => ref_,
        }
    }
}

impl Folder for Reads {
    type Output = iter::Once<ir::Stmt>;

    fn wrap_scope<R>(
        &mut self,
        ty: &ScopeTy,
        block: ir::Block,
        enter: impl FnOnce(&mut Self, ir::Block) -> R,
    ) -> R {
        if let ScopeTy::Toplevel = ty {
            visit_with(&block, |stmt: &ir::Stmt| match stmt {
                ir::Stmt::Expr {
                    target,
                    expr: ir::Expr::Read { source },
                } => match self.ssa_remappings.get(&source.weak()) {
                    Some(orig_ref) => {
                        self.ssa_remappings.insert(target.weak(), orig_ref.clone());
                    }
                    None => {
                        self.ssa_remappings.insert(target.weak(), source.clone());
                    }
                },
                _ => {}
            });
        }

        enter(self, block)
    }

    fn fold(&mut self, stmt: ir::Stmt) -> Self::Output {
        iter::once(match stmt {
            ir::Stmt::Expr { target, expr } => ir::Stmt::Expr {
                target,
                expr: match expr {
                    ir::Expr::Read { source } => ir::Expr::Read {
                        source: self.remap(source),
                    },
                    ir::Expr::ReadMember { obj, prop } => ir::Expr::ReadMember {
                        obj: self.remap(obj),
                        prop: self.remap(prop),
                    },
                    ir::Expr::Array { elems } => ir::Expr::Array {
                        elems: elems
                            .into_iter()
                            .map(|opt| opt.map(|(kind, ele)| (kind, self.remap(ele))))
                            .collect(),
                    },
                    ir::Expr::Object { props } => ir::Expr::Object {
                        props: props
                            .into_iter()
                            .map(|(kind, obj, prop)| (kind, self.remap(obj), self.remap(prop)))
                            .collect(),
                    },
                    ir::Expr::Unary { op, val } => ir::Expr::Unary {
                        op,
                        val: self.remap(val),
                    },
                    ir::Expr::Binary { op, left, right } => ir::Expr::Binary {
                        op,
                        left: self.remap(left),
                        right: self.remap(right),
                    },
                    ir::Expr::Delete { obj, prop } => ir::Expr::Delete {
                        obj: self.remap(obj),
                        prop: self.remap(prop),
                    },
                    ir::Expr::Yield { kind, val } => ir::Expr::Yield {
                        kind,
                        val: self.remap(val),
                    },
                    ir::Expr::Await { val } => ir::Expr::Await {
                        val: self.remap(val),
                    },
                    ir::Expr::Call { kind, func, args } => ir::Expr::Call {
                        kind,
                        func: self.remap(func),
                        args: args
                            .into_iter()
                            .map(|(kind, arg)| (kind, self.remap(arg)))
                            .collect(),
                    },
                    ir::Expr::Bool { .. }
                    | ir::Expr::Number { .. }
                    | ir::Expr::String { .. }
                    | ir::Expr::Null
                    | ir::Expr::Undefined
                    | ir::Expr::This
                    | ir::Expr::ReadMutable { .. }
                    | ir::Expr::ReadGlobal { .. }
                    | ir::Expr::RegExp { .. }
                    | ir::Expr::Function { .. }
                    | ir::Expr::CurrentFunction
                    | ir::Expr::Argument { .. } => expr,
                },
            },
            ir::Stmt::DeclareMutable { target, val } => ir::Stmt::DeclareMutable {
                target,
                val: self.remap(val),
            },
            ir::Stmt::WriteMutable { target, val } => ir::Stmt::WriteMutable {
                target,
                val: self.remap(val),
            },
            ir::Stmt::WriteGlobal { target, val } => ir::Stmt::WriteGlobal {
                target,
                val: self.remap(val),
            },
            ir::Stmt::WriteMember { obj, prop, val } => ir::Stmt::WriteMember {
                obj: self.remap(obj),
                prop: self.remap(prop),
                val: self.remap(val),
            },
            ir::Stmt::Return { val } => ir::Stmt::Return {
                val: self.remap(val),
            },
            ir::Stmt::Throw { val } => ir::Stmt::Throw {
                val: self.remap(val),
            },
            ir::Stmt::ForEach { kind, init, body } => ir::Stmt::ForEach {
                kind,
                init: self.remap(init),
                body,
            },
            ir::Stmt::IfElse { cond, cons, alt } => ir::Stmt::IfElse {
                cond: self.remap(cond),
                cons,
                alt,
            },
            ir::Stmt::Switch { discr, body } => ir::Stmt::Switch {
                discr: self.remap(discr),
                body,
            },
            ir::Stmt::Break { .. }
            | ir::Stmt::Continue { .. }
            | ir::Stmt::Debugger
            | ir::Stmt::Label { .. }
            | ir::Stmt::Loop { .. }
            | ir::Stmt::SwitchCase { .. }
            | ir::Stmt::Try { .. } => stmt,
        })
    }
}
