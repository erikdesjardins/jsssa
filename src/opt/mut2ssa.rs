use std::collections::{HashMap, HashSet};
use std::mem;

use crate::ir;
use crate::ir::traverse::{Folder, RunVisitor, ScopeTy, Visitor};

/// Converts read-only mutable vars to SSA, and removes write-only mutable vars.
///
/// Does not profit from multiple passes.
/// May profit from DCE running first; may create opportunities for DCE.
/// May create opportunities for read forwarding.
#[derive(Default)]
pub struct Mut2Ssa {
    mut_vars_to_replace: HashMap<ir::WeakRef<ir::Mut>, What>,
}

enum What {
    Convert(ir::Ref<ir::Ssa>),
    Remove,
}

#[derive(Default)]
struct CollectMutOpInfo<'a> {
    mut_ops: HashMap<&'a ir::Ref<ir::Mut>, State>,
    reads_in_scope: HashSet<&'a ir::Ref<ir::Mut>>,
    declared_at_toplevel_of_switch: HashSet<&'a ir::Ref<ir::Mut>>,
    at_toplevel_of_switch: bool,
    about_to_enter_switch: bool,
}

#[derive(PartialEq)]
enum State {
    ReadOnly { frozen: bool },
    WriteOnly,
    Invalid,
}

impl<'a> Visitor<'a> for CollectMutOpInfo<'a> {
    fn wrap_scope<R>(
        &mut self,
        ty: &ScopeTy,
        block: &'a ir::Block,
        enter: impl FnOnce(&mut Self, &'a ir::Block) -> R,
    ) -> R {
        match ty {
            ScopeTy::Function => {
                // functions are analyzed separately, since they can read ssa refs
                // before they are lexically declared
                let mut inner = Self::default();
                mem::swap(&mut inner.mut_ops, &mut self.mut_ops);
                let r = enter(&mut inner, block);
                mem::swap(&mut inner.mut_ops, &mut self.mut_ops);
                r
            }
            ScopeTy::Normal | ScopeTy::Toplevel | ScopeTy::Nonlinear => {
                let mut inner = Self::default();
                mem::swap(&mut inner.mut_ops, &mut self.mut_ops);
                mem::swap(&mut inner.reads_in_scope, &mut self.reads_in_scope);
                inner.at_toplevel_of_switch = self.about_to_enter_switch;
                let r = enter(&mut inner, block);
                mem::swap(&mut inner.mut_ops, &mut self.mut_ops);
                mem::swap(&mut inner.reads_in_scope, &mut self.reads_in_scope);
                self.about_to_enter_switch = false;
                r
            }
        }
    }

    fn visit(&mut self, stmt: &'a ir::Stmt) {
        match stmt {
            ir::Stmt::Expr {
                target: _,
                expr: ir::Expr::ReadMutable { source },
            } => {
                let our_state = State::ReadOnly { frozen: false };
                self.mut_ops
                    .entry(source)
                    .and_modify(|state| {
                        if state != &our_state {
                            *state = State::Invalid;
                        }
                    })
                    .or_insert(our_state);
                self.reads_in_scope.insert(source);
            }
            ir::Stmt::WriteMutable { target, val: _ } => {
                let our_state = State::WriteOnly;
                self.mut_ops
                    .entry(target)
                    .and_modify(|state| {
                        if state != &our_state {
                            *state = State::Invalid;
                        }
                    })
                    .or_insert(our_state);
            }
            ir::Stmt::DeclareMutable { target, val: _ } => {
                if self.reads_in_scope.contains(&target) {
                    // read before declaration
                    self.mut_ops.insert(target, State::Invalid);
                }
                if self.at_toplevel_of_switch {
                    self.declared_at_toplevel_of_switch.insert(target);
                }
            }
            ir::Stmt::Switch { .. } => {
                self.about_to_enter_switch = true;
            }
            ir::Stmt::SwitchCase { .. } => {
                for ref_ in self.declared_at_toplevel_of_switch.drain() {
                    match self.mut_ops.entry(ref_).or_insert(State::WriteOnly) {
                        State::ReadOnly { frozen } => {
                            *frozen = true;
                        }
                        State::WriteOnly | State::Invalid => {
                            // future reads already disallowed
                        }
                    }
                }
            }
            _ => {}
        }
    }
}

impl Folder for Mut2Ssa {
    type Output = Option<ir::Stmt>;

    fn wrap_scope<R>(
        &mut self,
        ty: &ScopeTy,
        block: ir::Block,
        enter: impl FnOnce(&mut Self, ir::Block) -> R,
    ) -> R {
        if let ScopeTy::Toplevel = ty {
            let mut collector = CollectMutOpInfo::default();
            collector.run_visitor(&block);
            self.mut_vars_to_replace = collector
                .mut_ops
                .into_iter()
                .flat_map(|(ref_, saw)| match saw {
                    State::ReadOnly { frozen: _ } => {
                        Some((ref_.weak(), What::Convert(ir::Ref::new(ref_.name_hint()))))
                    }
                    State::WriteOnly => Some((ref_.weak(), What::Remove)),
                    State::Invalid => None,
                })
                .collect();
        }

        enter(self, block)
    }

    fn fold(&mut self, stmt: ir::Stmt) -> Self::Output {
        match stmt {
            ir::Stmt::Expr {
                target,
                expr: ir::Expr::ReadMutable { source },
            } => match self.mut_vars_to_replace.get(&source.weak()) {
                Some(What::Convert(ssa_ref)) => Some(ir::Stmt::Expr {
                    target,
                    expr: ir::Expr::Read {
                        source: ssa_ref.clone(),
                    },
                }),
                Some(What::Remove) => unreachable!("removing mut read: {:?}", source),
                None => Some(ir::Stmt::Expr {
                    target,
                    expr: ir::Expr::ReadMutable { source },
                }),
            },
            ir::Stmt::DeclareMutable { target, val } => {
                match self.mut_vars_to_replace.get(&target.weak()) {
                    Some(What::Convert(ssa_ref)) => Some(ir::Stmt::Expr {
                        target: ssa_ref.clone(),
                        expr: ir::Expr::Read { source: val },
                    }),
                    Some(What::Remove) => None,
                    None => Some(ir::Stmt::DeclareMutable { target, val }),
                }
            }
            ir::Stmt::WriteMutable { target, val } => {
                match self.mut_vars_to_replace.get(&target.weak()) {
                    Some(What::Convert(_)) => unreachable!("converting mut write: {:?}", target),
                    Some(What::Remove) => None,
                    None => Some(ir::Stmt::WriteMutable { target, val }),
                }
            }
            _ => Some(stmt),
        }
    }
}
