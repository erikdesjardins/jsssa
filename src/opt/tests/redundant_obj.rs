use crate::opt::dce;
use crate::opt::forward;
use crate::opt::mut2ssa;
use crate::opt::redundant_obj;

macro_rules! passes {
    ( $cx:ident ) => {
        $cx.run::<mut2ssa::Mut2Ssa>("mut2ssa")
            .run::<forward::Reads>("forward-reads-redundancy")
            .converge::<dce::Dce>("dce-forwarded-reads")
            .run::<redundant_obj::LoadStore>("redundant-obj")
    };
}

case!(
    basic_read_to_read,
    |cx| passes!(cx),
    r#"
    let something = {};
    foo();
    g1 = something.x;
    g2 = something.x;
"#,
@r###"
something = {  }
_fun = <global foo>
<dead> = _fun()
_prp = "x"
_val = something[_prp]
<global g1> <- _val
<dead> = "x"
_val$1 = _val
<global g2> <- _val$1
"###);

case!(
    basic_write_to_read,
    |cx| passes!(cx),
    r#"
    let something = {};
    foo(something);
    something.x = x;
    g = something.x;
"#,
@r###"
something = {  }
_fun = <global foo>
<dead> = _fun(something)
_prp = "x"
_val = <global x>
something[_prp] <- _val
<dead> = "x"
_val$1 = _val
<global g> <- _val$1
"###);

case!(
    basic_write_to_write,
    |cx| passes!(cx),
    r#"
    let something = {};
    foo(something);
    something.x = x;
    something.x = y;
"#,
@r###"
something = {  }
_fun = <global foo>
<dead> = _fun(something)
<dead> = "x"
<dead> = <global x>
_prp = "x"
_val = <global y>
something[_prp] <- _val
"###);

case!(
    invalid_different_props,
    |cx| passes!(cx),
    r#"
    let something = {};
    foo();
    g1 = something.x;
    g2 = something.y;
"#,
@r###"
something = {  }
_fun = <global foo>
<dead> = _fun()
_prp = "x"
_val = something[_prp]
<global g1> <- _val
_prp$1 = "y"
_val$1 = something[_prp$1]
<global g2> <- _val$1
"###);

case!(
    invalidate_escape,
    |cx| passes!(cx),
    r#"
    let something = { x: 1 };
    let no_escape = { y: 2 };
    if (foo) {
        g = something;
    }
    g1 = something.x;
    g2 = no_escape.y; // forward
"#,
@r###"
_key = "x"
_val = 1
something = { [_key]: _val }
_key$1 = "y"
_val$1 = 2
<dead> = { [_key$1]: _val$1 }
_iff = <global foo>
<if> _iff:
    <global g> <- something
<else>:
    <empty>
_prp = "x"
_val$2 = something[_prp]
<global g1> <- _val$2
<dead> = "y"
_val$3 = _val$1
<global g2> <- _val$3
"###);

case!(
    invalidate_unknown_prop,
    |cx| passes!(cx),
    r#"
    let something = { x: 1 };
    let no_write = { y: 2 };
    if (foo) {
        something[foo] = 1;
        no_write[foo];
    }
    g1 = something.x;
    g2 = no_write.y; // forward
"#,
@r###"
_key = "x"
_val = 1
something = { [_key]: _val }
_key$1 = "y"
_val$1 = 2
no_write = { [_key$1]: _val$1 }
_iff = <global foo>
<if> _iff:
    _prp$1 = <global foo>
    _val$4 = 1
    something[_prp$1] <- _val$4
    _prp$2 = <global foo>
    <dead> = no_write[_prp$2]
<else>:
    <empty>
_prp = "x"
_val$2 = something[_prp]
<global g1> <- _val$2
<dead> = "y"
_val$3 = _val$1
<global g2> <- _val$3
"###);

case!(
    invalidate_inner_scope_writes,
    |cx| passes!(cx),
    r#"
    let obj = { foo: 1, bar: 2, not_written: 2 };
    // every ref should be forwarded except the two at the bottom
    if (something) {
        obj.foo;
        obj.foo = 3;
        obj.bar;
        if (something2) {
            obj.bar = 4;
        }
        obj.foo;
        obj.not_written;
    }
    obj.foo; // do not forward
    obj.bar; // do not forward
    obj.not_written;
"#,
@r###"
_key = "foo"
_val = 1
_key$1 = "bar"
_val$1 = 2
_key$2 = "not_written"
_val$2 = 2
obj = { [_key]: _val, [_key$1]: _val$1, [_key$2]: _val$2 }
_iff = <global something>
<if> _iff:
    <dead> = "foo"
    <dead> = _val
    _prp$2 = "foo"
    _val$3 = 3
    obj[_prp$2] <- _val$3
    <dead> = "bar"
    <dead> = _val$1
    _iff$1 = <global something2>
    <if> _iff$1:
        _prp$3 = "bar"
        _val$4 = 4
        obj[_prp$3] <- _val$4
    <else>:
        <empty>
    <dead> = "foo"
    <dead> = _val$3
    <dead> = "not_written"
    <dead> = _val$2
<else>:
    <empty>
_prp = "foo"
<dead> = obj[_prp]
_prp$1 = "bar"
<dead> = obj[_prp$1]
<dead> = "not_written"
<dead> = _val$2
"###);

case!(
    invalidate_calls,
    |cx| passes!(cx),
    r#"
    let obj = { foo: 1 };
    invalidate();
    obj.foo; // do not forward
    h = function() { return foo; };
"#,
@r###"
_key = "foo"
_val = 1
obj = { [_key]: _val }
_fun = <global invalidate>
<dead> = _fun()
_prp = "foo"
<dead> = obj[_prp]
_val$1 = <function>:
    _ret = <global foo>
    <return> _ret
<global h> <- _val$1
"###);

case!(
    invalidate_inner_scope_calls,
    |cx| passes!(cx),
    r#"
    let obj = { foo: 1 };
    while (x)
        invalidate();
    obj.foo; // do not forward
    h = function() { return foo; };
"#,
@r###"
_key = "foo"
_val = 1
obj = { [_key]: _val }
<loop>:
    _whl = <global x>
    <if> _whl:
        <empty>
    <else>:
        <break>
    _fun = <global invalidate>
    <dead> = _fun()
_prp = "foo"
<dead> = obj[_prp]
_val$1 = <function>:
    _ret = <global foo>
    <return> _ret
<global h> <- _val$1
"###);

case!(
    many_writes,
    |cx| passes!(cx),
    r#"
    let obj = { foo: 1 };
    obj.foo = 2;
    obj.foo = 3;
    obj.foo = 4;
    obj.foo = 5;
    g = obj;
"#,
@r###"
_key = "foo"
_val = 1
obj = { [_key]: _val }
<dead> = "foo"
<dead> = 2
<dead> = "foo"
<dead> = 3
<dead> = "foo"
<dead> = 4
_prp = "foo"
_val$1 = 5
obj[_prp] <- _val$1
<global g> <- obj
"###);

case!(
    reads_dont_propagate_to_parent,
    |cx| passes!(cx),
    r#"
    let obj = { foo: 1 };
    invalidate();
    if (bar) {
        g = obj.foo;
    } else {
        g = obj.foo;
    }
"#,
@r###"
_key = "foo"
_val = 1
obj = { [_key]: _val }
_fun = <global invalidate>
<dead> = _fun()
_iff = <global bar>
<if> _iff:
    _prp = "foo"
    _val$1 = obj[_prp]
    <global g> <- _val$1
<else>:
    _prp = "foo"
    _val$1 = obj[_prp]
    <global g> <- _val$1
"###);

case!(
    switch_invalidate_local,
    |cx| passes!(cx),
    r#"
    let a = { outer: 1 };
    switch (x) {
        case 1:
            a.inner = 2;
            a.inner;
        case 2:
            a.inner; // don't forward
    }
    a.outer;
"#,
@r###"
_key = "outer"
_val = 1
a = { [_key]: _val }
_swi = <global x>
_tst = 1
_tst$1 = 2
<switch> _swi:
    <case> _tst:
    _prp = "inner"
    _val$1 = 2
    a[_prp] <- _val$1
    <dead> = "inner"
    <dead> = _val$1
    <case> _tst$1:
    _prp$1 = "inner"
    <dead> = a[_prp$1]
<dead> = "outer"
<dead> = _val
"###);

case!(
    across_break_write,
    |cx| cx
        .run::<mut2ssa::Mut2Ssa>("mut2ssa")
        .run::<forward::Reads>("forward-reads-redundancy")
        .run::<redundant_obj::LoadStore>("redundant-obj"),
    r#"
    let o = { f: 0 };
    while (foo) {
        o.f += 1; // do not drop
        break;
        o.f = 3;
    }
"#,
@r###"
_key = "f"
_val = 0
o = { [_key]: _val }
<loop>:
    _whl = <global foo>
    <if> _whl:
        <empty>
    <else>:
        <break>
    <dead> = o
    _prp = "f"
    _lhs = o[_prp]
    _rhs = 1
    _val$1 = _lhs + _rhs
    o[_prp] <- _val$1
    <dead> = _val$1
    <break>
    <dead> = o
    _prp$1 = "f"
    _val$2 = 3
    o[_prp$1] <- _val$2
    <dead> = _val$2
"###);

case!(
    across_conditional_breaks_write,
    |cx| passes!(cx),
    r#"
    let o = { f: 0 };
    while (foo) {
        o.f += 1; // do not drop
        if (bar) {
            break;
        }
        o.f = 3;
    }
"#,
@r###"
_key = "f"
_val = 0
o = { [_key]: _val }
<loop>:
    _whl = <global foo>
    <if> _whl:
        <empty>
    <else>:
        <break>
    _prp = "f"
    _lhs = o[_prp]
    _rhs = 1
    _val$1 = _lhs + _rhs
    o[_prp] <- _val$1
    _iff = <global bar>
    <if> _iff:
        <break>
    <else>:
        <empty>
    _prp$1 = "f"
    _val$2 = 3
    o[_prp$1] <- _val$2
"###);

case!(
    across_deep_break_write,
    |cx| passes!(cx),
    r#"
    let o = { f: 0 };
    outer: while (foo) {
        o.f += 1; // do not drop
        while (bar) while (bar) {
            break outer;
        }
        o.f = 3;
    }
"#,
@r###"
_key = "f"
_val = 0
o = { [_key]: _val }
<label outer>:
    <loop>:
        _whl = <global foo>
        <if> _whl:
            <empty>
        <else>:
            <break>
        _prp = "f"
        _lhs = o[_prp]
        _rhs = 1
        _val$1 = _lhs + _rhs
        o[_prp] <- _val$1
        <loop>:
            _whl$1 = <global bar>
            <if> _whl$1:
                <empty>
            <else>:
                <break>
            <loop>:
                _whl$2 = <global bar>
                <if> _whl$2:
                    <empty>
                <else>:
                    <break>
                <break outer>
        _prp$1 = "f"
        _val$2 = 3
        o[_prp$1] <- _val$2
"###);

case!(
    across_conditional_breaks_read,
    |cx| passes!(cx),
    r#"
    let o = { f: 0 };
    while (foo) {
        o.f += 1;
        if (bar) {
            break;
        }
        g = o.f; // forward
    }
"#,
@r###"
_key = "f"
_val = 0
o = { [_key]: _val }
<loop>:
    _whl = <global foo>
    <if> _whl:
        <empty>
    <else>:
        <break>
    _prp = "f"
    _lhs = o[_prp]
    _rhs = 1
    _val$1 = _lhs + _rhs
    o[_prp] <- _val$1
    _iff = <global bar>
    <if> _iff:
        <break>
    <else>:
        <empty>
    <dead> = "f"
    _val$2 = _val$1
    <global g> <- _val$2
"###);

case!(
    cross_switch_read_invalidate,
    |cx| passes!(cx),
    r#"
    let o = { f: 0 };
    switch (bar) {
        case 1:
            o.f += 1; // do not forward
        case 2:
            g = o.f;
    }
"#,
@r###"
_key = "f"
_val = 0
o = { [_key]: _val }
_swi = <global bar>
_tst = 1
_tst$1 = 2
<switch> _swi:
    <case> _tst:
    _prp = "f"
    _lhs = o[_prp]
    _rhs = 1
    _val$1 = _lhs + _rhs
    o[_prp] <- _val$1
    <case> _tst$1:
    _prp$1 = "f"
    _val$2 = o[_prp$1]
    <global g> <- _val$2
"###);

case!(
    cross_switch_write_invalidate,
    |cx| passes!(cx),
    r#"
    let o = { f: 0 };
    switch (bar) {
        case 1:
            o.f += 1; // do not drop
        case 2:
            o.f = 3;
    }
"#,
@r###"
_key = "f"
_val = 0
o = { [_key]: _val }
_swi = <global bar>
_tst = 1
_tst$1 = 2
<switch> _swi:
    <case> _tst:
    _prp = "f"
    _lhs = o[_prp]
    _rhs = 1
    _val$1 = _lhs + _rhs
    o[_prp] <- _val$1
    <case> _tst$1:
    _prp$1 = "f"
    _val$2 = 3
    o[_prp$1] <- _val$2
"###);

case!(
    call_receiver,
    |cx| passes!(cx),
    r#"
    let something = { x: function() {} };
    something.x(); // receives `this`: do not opt
    let something2 = { x: function() {} };
    (0, something2.x)(); // does not receive `this`: opt
"#,
@r###"
_key = "x"
_val = <function>:
    <empty>
_obj = { [_key]: _val }
_prp = "x"
<dead> = _obj[_prp]()
_key$1 = "x"
_val$1 = <function>:
    <empty>
<dead> = { [_key$1]: _val$1 }
<dead> = "x"
_fun = _val$1
<dead> = _fun()
"###);
