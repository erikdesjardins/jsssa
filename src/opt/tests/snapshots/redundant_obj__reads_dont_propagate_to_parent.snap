---
created: "2019-04-10T03:59:56.999725400Z"
creator: insta@0.7.1
source: src/opt/tests/redundant_obj.rs
expression: "\n    let obj = { foo: 1 };\n    invalidate();\n    if (bar) {\n        g = obj.foo;\n    } else {\n        g = obj.foo;\n    }\n"
---
_key = "foo"
_val = 1
_ini = { [_key]: _val }
_fun = <global invalidate>
<dead> = _fun()
_iff = <global bar>
<if> _iff:
    _prp = "foo"
    _val$1 = _ini[_prp]
    <global g> <- _val$1
<else>:
    _prp = "foo"
    _val$1 = _ini[_prp]
    <global g> <- _val$1
